[SPARK-26538][SQL] Set default precision and scale for elements of postgres numeric array

a-shkarupin opened the pull request:
## What changes were proposed in this pull request?

When determining CatalystType for postgres columns with type `numeric[]` set the type of array element to `DecimalType(38, 18)` instead of `DecimalType(0,0)`. 

## How was this patch tested?

Tested with modified `org.apache.spark.sql.jdbc.JDBCSuite`.
Ran the `PostgresIntegrationSuite` manually.

a-shkarupin committed:
[SPARK-26538][SQL] Set default precision and scale for elements of postgres numeric array

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

mgaido91 reviewed (commented):
just a minor comment, otherwise seems reasonable. cc @gatorsmile @srowen for triggering the build when they are comfortable with the change.

dongjoon-hyun committed:
Add a test at PostgresIntegrationSuite

a-shkarupin committed:
[SPARK-26538][SQL] add a comment explaining the issue

a-shkarupin commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:
```diff
@@ -60,7 +60,8 @@ private object PostgresDialect extends JdbcDialect {
     case "timestamp" | "timestamptz" | "time" | "timetz" => Some(TimestampType)
     case "date" => Some(DateType)
-    case "numeric" | "decimal" => Some(DecimalType.bounded(precision, scale))
+    case "numeric" | "decimal" if precision != 0 => Some(DecimalType.bounded(precision, scale))
```
Updated per your suggestion.

a-shkarupin commented:
Would it make sense to add the tests from https://github.com/apache/spark/pull/23458/commits/74215de31cb68a5b6cf6c26ae0b6512926cdc150 as well ?

mgaido91 commented:
@a-shkarupin yes, I think so. cc @dongjoon-hyun who prepared that patch and can have other/better suggestions.

maropu commented:
ok to test

SparkQA commented:
**[Test build #100956 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100956/testReport)** for PR 23456 at commit [`31b0b04`](https://github.com/apache/spark/commit/31b0b046206a1f2316d69a083c8c38ad7b7b95ab).

maropu commented:
@a-shkarupin Please add the @dongjoon-hyun 's test in `PostgresIntegrationSuite.scala`?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6843/
Test PASSed.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:
```diff
@@ -60,7 +60,12 @@ private object PostgresDialect extends JdbcDialect {
     case "timestamp" | "timestamptz" | "time" | "timetz" => Some(TimestampType)
     case "date" => Some(DateType)
-    case "numeric" | "decimal" => Some(DecimalType.bounded(precision, scale))
+    case "numeric" | "decimal" => if (precision > 0) {
```
I would like to confirm just in case; we don't check `scale` in this pr? Probably, this might be related to the discussion: https://github.com/apache/spark/pull/23458#discussion_r245480257

a-shkarupin replied:
Postgres [doc](https://www.postgresql.org/docs/9.6/datatype-numeric.html) says that 

> The precision must be positive, the scale zero or positive. 

What the postgres jdbc driver returned in case of `numeric` was 0 for both scale and precision.
The condition proposed in the linked ticket and currently used [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L230) was roughly `precision > 0 || scale > 0`, but I can not come up with a valid case having precision <=0 while having scale > 0. 
Is there another case where we would have a decimal with precision 0?
Could someone explain?

maropu replied:
Yea, but, I think we'd be better to add the check `scale > 0`, too, just for safeguards.

mgaido91 replied:
I think this is fine actually,. We do not support decimals with precision < 0, so this is most likely enough.

dongjoon-hyun replied:
Actually, I still agree with @maropu (https://github.com/apache/spark/pull/23456#discussion_r), but it looks okay because this is `PostgresDialect.scala`.

mgaido91 replied:
if we'd add `|| scale > 0`, we'd allow a percision <= 0,which doesn't make any sense and it is not supported by Spark's decimal. So I think this is fine.

maropu commented:
nit: Could you clean up the title (plz move `…stgres numeric array` into the title)?

SparkQA commented:
**[Test build #100956 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100956/testReport)** for PR 23456 at commit [`31b0b04`](https://github.com/apache/spark/commit/31b0b046206a1f2316d69a083c8c38ad7b7b95ab).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100956/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6859/
Test PASSed.

SparkQA commented:
**[Test build #100976 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100976/testReport)** for PR 23456 at commit [`77bbcb5`](https://github.com/apache/spark/commit/77bbcb540f9f43b5aa9334e052c4e887460eec41).

a-shkarupin commented:
> @a-shkarupin Please add the @dongjoon-hyun 's test in `PostgresIntegrationSuite.scala`? Also, since Jenkins don't run the test, please check if the test passed in your env?

Added the test.

Ran tests as follows:
```
./build/mvn install -DskipTests
./build/mvn test -Pdocker-integration-tests -pl :spark-docker-integration-tests_2.12
```
Got following result:
```
Run completed in 3 minutes, 36 seconds.
Total number of tests run: 21
Suites: completed 5, aborted 0
Tests: succeeded 21, failed 0, canceled 0, ignored 5, pending 0
All tests passed.
```

> nit: Could you clean up the title (plz move …stgres numeric array into the title)?

Cleaned up.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100976/
Test PASSed.

maropu commented:
Can you update the PR description (`How was this patch tested?`) ?

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala:
```diff
@@ -60,7 +60,12 @@ private object PostgresDialect extends JdbcDialect {
+    } else {
+      // SPARK-26538: handle numeric without explicit precision and scale.
+      Some(DecimalType. SYSTEM_DEFAULT)
+    }
```
Hi, @a-shkarupin .
Could you follow the existing style?
- https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L215-L217

a-shkarupin replied:
Updated per your suggestion, kept the comment as suggested [here](https://github.com/apache/spark/pull/23456#discussion_r245425442).

a-shkarupin committed:
[SPARK-26538][SQL] follow existing style

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6905/
Test PASSed.

SparkQA commented:
**[Test build #101030 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101030/testReport)** for PR 23456 at commit [`c72e214`](https://github.com/apache/spark/commit/c72e2149bc421019692bf028784d2bd511b9e11c).

a-shkarupin commented:
> Can you update the PR description (`How was this patch tested?`) ?

Updated.

SparkQA commented:
**[Test build #101030 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101030/testReport)** for PR 23456 at commit [`c72e214`](https://github.com/apache/spark/commit/c72e2149bc421019692bf028784d2bd511b9e11c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101030/
Test PASSed.

maropu commented:
LGTM

mgaido91 commented:
LGTM too, thanks.

dongjoon-hyun reviewed (approved):
+1, LGTM

dongjoon-hyun commented:
Thank you all! Merged to `master`.

dongjoon-hyun commented:
Hi, @a-shkarupin . What is your Apache JIRA id?
I'm trying to add you to Apache Spark contributor group.

a-shkarupin commented:
Hi @dongjoon-hyun . My Apache JIRA username is alsh. I reported SPARK-26538.
Thanks.

dongjoon-hyun commented:
HI, @a-shkarupin . Yep. `alsh` is added to Spark contributor group and SPARK-26538 is assigned to you. If you are not in the contributor group, we cannot assign you an issue. Since you are added now, there is no problem in assigning.
